Progress for week one of the dashboard exercise

NilsHaeuslerEso opened the pull request:
see [week1](https://github.com/ui-bootcamp/bootcamp-schedule/blob/master/angular/dashboard/README.md#week-1) for the tasks

florianwarnkeeso committed:
feat(context): added theming and angular material

florianwarnkeeso committed:
feat: added media backend (search missing)

NilsHaeuslerEso committed:
feat: basic news module (front- and backend)

NilsHaeuslerEso committed:
feat: add grid view for dashboard apps

florianwarnkeeso committed:
Feature/flwa6204/search service (#4)

feat: add search

NilsHaeuslerEso committed:
feat: media colum view and fake player

NilsHaeuslerEso committed:
feat: page navigation via the router

florianwarnkeeso committed:
Added user service and core module

* feat: add user service

* feat: add favorites to media app, added core modul

NilsHaeuslerEso committed:
feat: fake player supports now all relevant tasks

NilsHaeuslerEso committed:
feat: load album cover of current track

NilsHaeuslerEso committed:
refactor: media selection is now a seperate component

NilsHaeuslerEso committed:
feat: local search in dashboard apps

florianwarnkeeso committed:
feat: added search result action

* feat: added search result action

* feat: optimized wrong api return

* feat: optimized wrong api return

NilsHaeuslerEso committed:
feat: the album comver is now shown in the search results

florianwarnkeeso committed:
feat: added unit tests

florianwarnkeeso committed:
refactor: refactoring of css/html

NilsHaeuslerEso committed:
docs: updated overview

florianwarnkeeso committed:
refactor: make the article component to a presentation component

florianwarnkeeso committed:
refactor: reworked tests

NilsHaeuslerEso committed:
refactor: change description of test to be consistent

NilsHaeuslerEso committed:
refactor: change project structure

florianwarnkeeso committed:
refactor: renaming, add unsubscribe & cleanup searchModel

florianwarnkeeso committed:
refactor: added smart component for fake player

NilsHaeuslerEso committed:
refactor: media-selection into smart and dump components

Commit:
refactor: renaming

NilsHaeuslerEso committed:
refactor: add the type to the data models

NilsHaeuslerEso committed:
refactor: use model type for favorites query

NilsHaeuslerEso committed:
refactor: remove the SearchResult model

NilsHaeuslerEso committed:
refactor: use proper calls to bootcamp api

NilsHaeuslerEso committed:
refactor: remove the SearchResultType in favor of type gurads

Commit:
feat: change favorite status via long press directive

olie4350 commented on src/app/app.component.ts:
```diff
@@ -1,10 +1,28 @@
 })
 export class AppComponent {
-  title = 'dashboard';
+  public title = 'dashboard';
```
readonly

olie4350 commented on src/app/app.component.ts:
```diff
@@ -1,10 +1,28 @@
+  public isDarkTheme$: Observable<boolean>;
+
+  constructor(
+    private themeService: ThemeService,
```
readonly

olie4350 commented on src/app/app.component.ts:
```diff
@@ -1,10 +1,28 @@
+    overlayContainer: OverlayContainer
+  ) {
+    this.isDarkTheme$ = this.themeService.isDarkTheme$;
+    this.isDarkTheme$.subscribe((value: boolean) => {
```
Subscription merken und bei ngOnDestroy unsubscribe() aufrufen

inkue replied:
Subscriptions immer merken und im ngOnDestroy unsubscriben

olie4350 commented on src/app/dashboard/dashboard.component.ts:
```diff
@@ -0,0 +1,22 @@
+export class DashboardComponent {
+  public isDarkTheme$: Observable<boolean>;
+
+  constructor(private themeService: ThemeService) {
```
readonly

olie4350 commented on src/app/dashboard/dashboard.component.ts:
```diff
@@ -0,0 +1,22 @@
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class DashboardComponent {
+  public isDarkTheme$: Observable<boolean>;
```
readonly

olie4350 commented on src/app/dashboard/media/components/fake-player-host/fake-player-host.component.html:
```diff
@@ -0,0 +1,2 @@
+<db-fake-player [track]="track" [albumCover]="albumCover | async">
```
albumCover$

olie4350 commented on src/app/dashboard/media/components/fake-player-host/fake-player-host.component.ts:
```diff
@@ -0,0 +1,42 @@
+    private cd: ChangeDetectorRef
+  ) {
+    this.albumCover = new Observable<string>();
+    this.subscription = this.playlistService.nextTrack$.subscribe(nextTrack => {
```
Typ bei nextTrack nicht vergessen

olie4350 replied:
Ich würde hier 2 Observables machen. 
`this.albumCover$ = this.playlistService.nextTrack$.pipe(
switchMap(track => this.mediaService.getAlbum$(nextTrack.albumId)),
map(album => album.coverSmall)
);`
Das könntet ihr auch direkt bei der Deklaration machen
`public readonly albumCover$: Observable<string> = ...`

Und dann noch ein Observable um track zu initialisieren.

inkue commented on src/app/app.component.html:
```diff
@@ -1 +1,6 @@
-<h1>{{title}} works!</h1>
+<div
+  [ngClass]="{ 'dark-theme': isDarkTheme$ | async }"
```
Wir würden bei uns in Navi zuerst class= und dann danach [ngClass]= schreiben, würde ich hier aber nicht als dringende Änderung ansehen

olie4350 commented on src/app/dashboard/media/components/fake-player-host/fake-player-host.component.ts:
```diff
@@ -0,0 +1,42 @@
+    private playlistService: PlaylistService,
+    private cd: ChangeDetectorRef
+  ) {
+    this.albumCover = new Observable<string>();
```
Mit 2 Observables wird diese Initialisierung nicht mehr benötigt.

olie4350 commented on src/app/dashboard/media/components/fake-player-host/fake-player-host.component.ts:
```diff
@@ -0,0 +1,42 @@
+      this.track = nextTrack;
+      this.albumCover = this.mediaService
+        .getAlbum(nextTrack.albumId)
+        .pipe(map(album => album.coverSmall));
```
Typen nicht vergessen

olie4350 commented on src/app/dashboard/media/components/fake-player/fake-player.component.html:
```diff
@@ -0,0 +1,16 @@
+<div class="fake-player" *ngIf="track && track.title && track.duration">
```
am besten in 2 separaten Zeilen

olie4350 replied:
Hier sollte man lieber mit Elvis-Operator auf undefined absichern
`*ngIf="track?.title && track?.duration"`

inkue replied:
Bei uns wäre die Reihenfolge:
*ngIf=""
class=""

olie4350 commented on src/app/dashboard/media/components/fake-player/fake-player.component.html:
```diff
@@ -0,0 +1,16 @@
+  <img [src]="albumCover" alt="Album Cover" />
+  <div class="fake-player-wrapper">
+    <b>{{ track.title }}</b>
+    <br />
```
Lieber padding/margin in scss nutzen

olie4350 commented on src/app/dashboard/media/components/fake-player/fake-player.component.html:
```diff
@@ -0,0 +1,16 @@
+      <mat-icon>skip_previous</mat-icon>
+      <mat-icon (click)="onPlayPauseClick()">{{ playPauseIcon }}</mat-icon>
+      <mat-icon>skip_next</mat-icon>
+      <div>{{ track.duration * 0.4 | secondsToMinutes }}</div>
```
keine magic numbers -> Konstante in ts-Datei definieren

olie4350 commented on src/app/dashboard/media/components/fake-player/fake-player.component.scss:
```diff
@@ -0,0 +1,28 @@
+  justify-content: center;
+}
+
+mat-progress-bar {
```
besser eine Klasse definieren

inkue commented on src/app/dashboard/media/components/fake-player-host/fake-player-host.component.ts:
```diff
@@ -0,0 +1,42 @@
+})
+export class FakePlayerHostComponent implements OnDestroy {
+  public track: Track | undefined;
+  public albumCover: Observable<string>;
```
albumCover$ mit $ am Ende, weil Observable :) dann sieht man auch im HTML gleich, dass ein Observable genutzt wurde

olie4350 commented on src/app/dashboard/media/components/fake-player/fake-player.component.ts:
```diff
@@ -0,0 +1,25 @@
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class FakePlayerComponent {
+  @Input() track: Track | undefined;
```
Eingentlich sollte hier kein undefined möglich sein.

olie4350 commented on src/app/dashboard/media/components/fake-player/pipes/seconds-to-minutes.pipe.spec.ts:
```diff
@@ -0,0 +1,18 @@
+import { SecondsToMinutesPipe } from './seconds-to-minutes.pipe';
+
+describe('SecondsToMinutesPipe', () => {
```
nice :)

inkue commented on src/app/dashboard/media/components/media-selection/media-selection.component.html:
```diff
@@ -0,0 +1,55 @@
+    (selectionChange)="onSelectedArtistChanged($event.option)"
+  >
+    <mat-list-option
+      *ngFor="let artist of artists"
```
Würden wir in Navi so sortieren:

```
*ngFor="let artist of artists"
dbFavorite
[value]="artist"
[isFavorite]="artist.isFavorite"
(dblclick)="toggleFavorite(artist)"
```

olie4350 replied:
bitte alle Direktiven, Inputs, Outputs zusammenmachen -> Sortierung.

olie4350 commented on src/app/dashboard/media/components/media-selection-host/media-selection-host.component.ts:
```diff
@@ -0,0 +1,51 @@
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class MediaSelectionHostComponent implements OnInit {
+  public artists$ = new Observable<Artist[]>();
```
unnötige Initialisierung

olie4350 commented on src/app/dashboard/media/components/media-selection-host/media-selection-host.component.ts:
```diff
@@ -0,0 +1,51 @@
+  }
+
+  public onSelectedArtistChanged(artist: Artist): void {
+    this.albums$ = this.mediaService
```
Hier würde ich mit Subject arbeiten. Und dann einfach den Wert in next() übergeben.

inkue commented on src/app/dashboard/media/media.component.ts:
```diff
@@ -0,0 +1,33 @@
+  }
+
+  public onToggleFullscreen(): void {
+    if (this.router.url.indexOf('/media') > -1) {
```
Könnte man hier über this.isFullscreen gehen statt noch ein zweites Mal die URL zu checken?

florianwarnkeeso replied:
nein

olie4350 commented on src/app/dashboard/media/components/media-selection/media-selection.component.scss:
```diff
@@ -0,0 +1,21 @@
+  display: flex;
+}
+
+mat-selection-list {
```
Hier ist genauso, am besten die Klassen definieren und die stylen

olie4350 commented on src/app/dashboard/media/components/media-selection/media-selection.component.ts:
```diff
@@ -0,0 +1,53 @@
+  public onSelectedArtistChanged(option: MatListOption): void {
+    option.selectionList.deselectAll();
+    option.selected = true;
+    this.selectedArtistChanged.emit(option.value);
```
nice

olie4350 commented on src/app/dashboard/media/media.component.ts:
```diff
@@ -0,0 +1,33 @@
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class MediaComponent {
+  public isFullscreen: boolean;
```
readonly

olie4350 commented on src/app/dashboard/media/models/album.model.ts:
```diff
@@ -0,0 +1,24 @@
+import { Artist } from './artist.model';
+import { Article } from '../../news/models/article.model';
+
+export class Album {
```
man konnte auch einfach als Interface machen

olie4350 commented on src/app/dashboard/media/shared/media.service.ts:
```diff
@@ -0,0 +1,123 @@
+
+  constructor(
+    private http: HttpClient,
+    private storageService: StorageService
```
readonly

olie4350 commented on src/app/dashboard/media/shared/media.service.ts:
```diff
@@ -0,0 +1,123 @@
+    private storageService: StorageService
+  ) {}
+
+  public getArtist(id: number): Observable<Artist> {
```
$ fehlt

olie4350 commented on src/app/dashboard/media/shared/media.service.ts:
```diff
@@ -0,0 +1,123 @@
+  public getArtist(id: number): Observable<Artist> {
+    const artistURL = `${this.artistsURL}${id}`;
+    return this.http.get<Artist>(artistURL).pipe(
+      map(artist => {
```
Typ fehlt

olie4350 commented on src/app/dashboard/media/shared/media.service.ts:
```diff
@@ -0,0 +1,123 @@
+    );
+  }
+
+  public getAlbum(id: number): Observable<Album> {
```
$

olie4350 commented on src/app/dashboard/media/shared/media.service.ts:
```diff
@@ -0,0 +1,123 @@
+  providedIn: 'root'
+})
+export class MediaService {
+  private artistsURL = `${environment.baseUrl}artists/`;
```
Die Konstanten sollte man auch als Konstanten deklarieren.
`private static readonly ARTISTS_URL = ...` oder ganz oben als `const ARTISTS_URL = ...`

olie4350 commented on src/app/dashboard/media/shared/media.service.ts:
```diff
@@ -0,0 +1,123 @@
+    );
+  }
+
+  public getAllArtists(): Observable<Artist[]> {
```
$

olie4350 commented on src/app/dashboard/news/components/article/article.component.html:
```diff
@@ -0,0 +1,16 @@
+<mat-card
+  dbFavorite
+  [isFavorite]="article.isFavorite"
```
article?.isFavorite

olie4350 commented on src/app/dashboard/news/components/article/article.component.html:
```diff
@@ -0,0 +1,16 @@
+  [isFavorite]="article.isFavorite"
+  (dblclick)="dblClickArticle(article)"
+>
+  <mat-card-title [innerText]="article.title"></mat-card-title>
```
article?.title

olie4350 commented on src/app/dashboard/news/components/article/article.component.ts:
```diff
@@ -0,0 +1,23 @@
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class ArticleComponent {
+  @Input() public article!: Article;
```
wofür "!"?

olie4350 commented on src/app/dashboard/news/components/article/article.component.ts:
```diff
@@ -0,0 +1,23 @@
+  @Output() public readonly toggleArticleFavorite = new EventEmitter<Article>();
+
+  public dblClickArticle(article: Article): void {
+    this.toggleArticleFavorite.emit(article);
```
Wenn eine Methode nur aus einer Zeile besteht und kein Logging benötigt wird, würde ich direkt im Template machen

olie4350 commented on src/app/dashboard/news/news.component.html:
```diff
@@ -0,0 +1,23 @@
+      <span class="spacer"></span>
+      <db-search *ngIf="isFullscreen" searchScope="news"></db-search>
+      <button mat-button (click)="onToggleFullscreen()">
+        <mat-icon *ngIf="isFullscreen">fullscreen_exit</mat-icon>
```
Prinzipiell ok, aber da unser Texttool im Projekt mit "ngIf=true/false" nicht klar kommt, sollte man mit ng-template lösen.
`<mat-icon *ngIf="isFullscreen; else noFullScreen">fullscreen_exit</mat-icon>`
`<ng-template><mat-icon #noFullScreen>fullscreen</mat-icon></ng-template>`

olie4350 commented on src/app/dashboard/news/news.component.ts:
```diff
@@ -0,0 +1,44 @@
+  public isFullscreen: boolean;
+
+  constructor(
+    private newsService: NewsService,
```
readonly

olie4350 commented on src/app/dashboard/news/news.component.ts:
```diff
@@ -0,0 +1,44 @@
+    private router: Router,
+    private storageService: StorageService
+  ) {
+    this.articles$ = new Observable<Article[]>();
```
Keine Initialisierung wird benötigt

olie4350 commented on src/app/dashboard/news/shared/news.service.spec.ts:
```diff
@@ -0,0 +1,82 @@
+import { Article } from '../models/article.model';
+
+describe('NewsService', () => {
+  let injector: TestBed;
```
ts-mockito verwenden

olie4350 commented on src/app/dashboard/news/shared/news.service.ts:
```diff
@@ -0,0 +1,42 @@
+  providedIn: 'root'
+})
+export class NewsService {
+  private URL = `${environment.baseUrl}news`;
```
als Konstante deklarieren

olie4350 commented on src/app/dashboard/news/shared/news.service.ts:
```diff
@@ -0,0 +1,42 @@
+    private storageService: StorageService
+  ) {}
+
+  public getAllArticles(): Observable<Article[]> {
```
$

olie4350 commented on src/app/dashboard/news/shared/news.service.ts:
```diff
@@ -0,0 +1,42 @@
+    return this.getAllArticles().pipe(
+      map(articles =>
+        articles.filter(
+          x =>
```
bad 👎 
(article: Article)

olie4350 commented on src/app/dashboard/shared/core/directives/favorite.directive.ts:
```diff
@@ -0,0 +1,26 @@
+  selector: '[dbFavorite]'
+})
+export class FavoriteDirective implements AfterContentChecked {
+  @Input() isFavorite: boolean;
```
Man kann auch hier direkt initialisieren als Default Wert

olie4350 commented on src/app/dashboard/shared/core/directives/favorite.directive.ts:
```diff
@@ -0,0 +1,26 @@
+    this.isFavorite = false;
+  }
+
+  ngAfterContentChecked(): void {
```
public

olie4350 commented on src/app/dashboard/shared/core/pipes/favorite-icon.pipe.ts:
```diff
@@ -0,0 +1,14 @@
+})
+export class FavoriteIconPipe implements PipeTransform {
+  transform(favorite: boolean): string {
+    if (favorite) {
```
Besser: `return favorite ? 'favorite' : ''`

olie4350 commented on src/app/dashboard/shared/search/search.component.html:
```diff
@@ -0,0 +1,33 @@
+  <mat-autocomplete #auto="matAutocomplete">
+    <mat-option
+      *ngFor="let result of searchResults$ | async"
+      [value]="searchField.value"
```
Sortierung!

olie4350 commented on src/app/dashboard/shared/search/search.component.html:
```diff
@@ -0,0 +1,33 @@
+      [isFavorite]="result.isFavorite"
+      (onSelectionChange)="onSearchResultSelected(result)"
+    >
+      <mat-icon *ngIf="!result.tracks; else image">{{
```
nice 👍

olie4350 commented on src/app/dashboard/shared/search/search.component.ts:
```diff
@@ -0,0 +1,106 @@
+    (Artist | Album | Track | Article)[]
+  >();
+  public lastUserSearches = '';
+  private subscriptions: Subscription[] = [];
```
private readonly subscriptions: Subscription = new Subscription();

inkue commented on src/app/dashboard/news/news.component.ts:
```diff
@@ -0,0 +1,44 @@
+    this.articles$ = this.newsService.getAllArticles();
+  }
+  public onToggleFullscreen(): void {
+    if (this.router.url.indexOf('/news') > -1) {
```
Hier auch: Könnte man einfach this.isFullscreen verwenden statt noch mal die URL zu checken?

olie4350 commented on src/app/dashboard/shared/search/search.component.ts:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  public ngOnDestroy(): void {
+    this.subscriptions.forEach(subscription => subscription.unsubscribe());
```
es würde schon reichen: `this.subscriptions.unsubscribe();`

olie4350 commented on src/app/dashboard/shared/search/search.component.ts:
```diff
@@ -0,0 +1,106 @@
+        element.scrollIntoView();
+      }
+    } else if (Track.isTrack(result)) {
+      this.subscriptions.push(
```
man kann eine Methode add() auch verwenden

olie4350 commented on src/app/dashboard/shared/search/search.component.ts:
```diff
@@ -0,0 +1,106 @@
+export class SearchComponent implements OnInit, OnDestroy {
+  @Input() searchScope = 'global';
+  public searchField: FormControl = new FormControl();
+  public searchResults$ = new Observable<
```
keine Initialisierung + readonly

florianwarnkeeso committed:
feat: add lock directive

olie4350 commented on src/app/dashboard/shared/search/search.component.ts:
```diff
@@ -0,0 +1,106 @@
+
+  private displaySearchResult(result: Artist | Album | Track | Article): void {
+    if (Article.isArticle(result)) {
+      const element = document.getElementById('article' + result.id);
```
Typ fehlt

olie4350 commented on src/app/dashboard/shared/search/search.component.ts:
```diff
@@ -0,0 +1,106 @@
+        this.mediaService
+          .getArtist(result.id)
+          .subscribe((artist: Artist) =>
+            this.playlistService.queueTrack(artist.albums[0].tracks[0])
```
Ich würde auf undefined prüfen

olie4350 commented on src/app/dashboard/shared/search/shared/search.service.spec.ts:
```diff
@@ -0,0 +1,75 @@
+    mockNewsService = mock(NewsService);
+    mockStorageService = mock(StorageService);
+    when(mockMediaService.getAllTracksWhichContain(anything())).thenReturn(
+      of([new Track(9, 9, 9, 'THE TRACK', 123, '', '')])
```
vllt besser als Konstanten oben definieren

olie4350 commented on src/app/dashboard/shared/search/shared/search.service.ts:
```diff
@@ -0,0 +1,50 @@
+    searchTerm: string,
+    scope: string
+  ): Observable<(Artist | Album | Track | Article)[]> {
+    const sources: Observable<Artist[] | Album[] | Track[] | Article[]>[] = [];
```
$

olie4350 commented on src/app/dashboard/shared/search/shared/search.service.ts:
```diff
@@ -0,0 +1,50 @@
+    scope: string
+  ): Observable<(Artist | Album | Track | Article)[]> {
+    const sources: Observable<Artist[] | Album[] | Track[] | Article[]>[] = [];
+    if (scope === 'media' || scope === 'global') {
```
Scope als Enum

inkue commented on src/app/dashboard/shared/core/directives/favorite.directive.ts:
```diff
@@ -0,0 +1,26 @@
+  @Input() isFavorite: boolean;
+
+  constructor(private el: ElementRef, private renderer: Renderer2) {
+    this.isFavorite = false;
```
Was genau ist der Hintergrund? Es soll initial auf false stehen, sofern man keinen Input reingibt?
Dann einfach beim Input:

> @Input() isFavorite: boolean = false;

olie4350 commented on src/app/dashboard/shared/services/playlist.service.spec.ts:
```diff
@@ -0,0 +1,31 @@
+
+describe('PlaylistService', () => {
+  let playlistService: PlaylistService;
+  let dummyTrack: Track;
```
als Konstante

olie4350 commented on src/app/dashboard/shared/services/storage.service.spec.ts:
```diff
@@ -0,0 +1,147 @@
+    spyOn(localStorage, 'clear').and.callFake(mockLocalStorage.clear);
+  });
+
+  describe('getLastQueries', () => {
```
nice 👍

florianwarnkeeso committed:
refactor: refactored hostbinding

inkue commented on src/app/dashboard/shared/services/playlist.service.ts:
```diff
@@ -0,0 +1,17 @@
+  private nextTrack$$ = new Subject<Track>();
+  public nextTrack$: Observable<Track> = this.nextTrack$$;
+
+  constructor() {}
```
Kann man weglassen, wenn leer :)

inkue reviewed (changes requested):
# Review

## Continuous Integration

- [x] Travis configuration is complete
  - production build
  - test
  - lint
- [x] Build for master branch is always green
- [x] Travis badge is set in README

## Konventionen + Setup

- [x] Prefix `db` is used
- [x] `scss` is used
- [x] commitlint is configured & the right messages are used
  > Hint: Provide a scope to clarify which part of your app got and awesome update: `feat(media): add favorites list`
- [x] prettier is configured (e.g. with `.prettierrc`)
- [x] Working with Branches (Features, Bugfixes, Refactoring, etc.)
- [x] Pull requests are used to integrate features into master
- [x] Pull requests can only be merged into the master by mutual code review
- [x] Try to avoid merge commits. The git history should ideally be completely linear in the end (works with "git rebase" instead of merge)

## Acceptance Critiria

### General

- [x] The dashboard contains the following apps: Media and News
- [x] Media app has a dashboard view
- [x] Media app has a full-screen view
- [x] Media app can be searched
- [x] News app has a dashboard view
- [x] News app has a full-screen view
- [x] News app can be searched

### Dashboard

- [x] The dashboard should get a uniform theme / UI for all apps (also future ones)
- [x] You should be able to switch between a light / dark theme (or two themes of your choice)
- [x] The dashboard can be completely operated in touch mode (except drag & drop)
- [x] All components use ChangeDetectionStrategy.OnPush (tv Change Detection Explained)
- [x] The dashboard offers a global search that includes individual app searches
- [ ] If the search result is clicked, it will be displayed in the respective app
So halb ;) Der Track wird abgespielt, News wurde mir jetzt nicht geöffnet angezeigt o.Ä.

### News

- [x] Internal search only gives News search results

### Media

- [x] Internal search only gives Media search results
- [x] It has a full text search that searches for artist, albums and tracks
- [x] Output of the search includes meta-information, as well as possible pictures of the artist / album.
- [ ] Artists and tracks can be marked as favorites
Also irgendwie hab ich die Funktion nicht gefunden, obwohls im Code ja drin war? xD
- [x] Favorites werden in LocalStorage oder IndexedDB gespeichert
- [ ] Favorites should be prominently highlighted in the Media App and in the search
Konnte ich mangels Favorite-Button finden nicht testen :D
- [x] fake player should be included to play tracks, or playlists (albums / search results).



Wenn ich das grad richtig im Kopf hab vom Code anschauen, habt ihr bisher keine der Direktiven und folgenden Sachen implementiert, oder? Aber war ja auch nur der Teil von einer Woche.
Kann man sich ggfs. bei einem weiteren PR noch anschauen, falls ihr das noch umsetzt.

### Directives

- [ ] Long Press Directive implemented
- [ ] Long Press Directive used to set favorites
- [ ] UI States Directive implemented
- [ ] UI States Directive used for all clickable items
- [ ] Locking Directive:
  - [ ] lists not scrollable
  - [ ] important buttons grayed out (e.g. media player)
  - [ ] images hidden
  - [ ] Bonus: multi stage blocking

### Outside Angular

- [ ] Every dev should use ousideAngular
  - [ ] Dev 1
  - [ ] Dev 2
- [ ] Drag & Drop tiles
- [ ] timeline in fake player
- [ ] sidebar
  - [ ] Bonus: animated return

## Bonus

- [ ] Geo Location
- [ ] POI Search
- [ ] Weather App

## Tests

- [x] Jest is used

## Project Structure

### Directory Structure

- `containers`: contains smart/container components
- `components`: contains dumb/presentation components
Hätte man hier evtl. stärker machen können

## Code

See code comments of this pull request.

florianwarnkeeso committed:
feat: add draggable sidebar

NilsHaeuslerEso committed:
feat: movable progress bar is animated outside angular

NilsHaeuslerEso committed:
feat: map with poi radius search

feat: map with agm

feat: search for nearby places

feat: toggle favorite places

feat: added maps

florianwarnkeeso committed:
feat: add weather app

florianwarnkeeso committed:
refactor: add spyOn Date, to get tests running

florianwarnkeeso committed:
refactor: use viewchild instead of document.getelementbyid

florianwarnkeeso committed:
refactor: apply coding guidelines

florianwarnkeeso committed:
refactor: fixed tests

florianwarnkeeso committed:
refactor: remove apikey
